Description: If vendor utilities are found *and* this driver doesn't support RR1.2,
             offer fallback to those vendor utilities instead.
             This helps to alleviate possible crashes from unsupported features in RR1.0
             drivers.
Bug: http://bugzilla.gnome.org/show_bug.cgi?id=571214
Bug-Ubuntu: https://bugs.edge.launchpad.net/ubuntu/+source/gnome-control-center/+bug/319725
Patch: http://bugzilla.gnome.org/attachment.cgi?id=128398&action=view

Index: gnome-control-center-2.31.90/capplets/display/display-capplet.ui
===================================================================
--- gnome-control-center-2.31.90.orig/capplets/display/display-capplet.ui	2010-08-17 02:46:46.000000000 +1000
+++ gnome-control-center-2.31.90/capplets/display/display-capplet.ui	2010-09-01 17:42:34.192617001 +1000
@@ -414,6 +414,34 @@
       <action-widget response="-7">button2</action-widget>
     </action-widgets>
   </object>
+  <object class="GtkMessageDialog" id="legacy_question">
+    <property name="border_width">5</property>
+    <property name="resizable">False</property>
+    <property name="window_position">GTK_WIN_POS_CENTER_ON_PARENT</property>
+    <property name="type_hint">GDK_WINDOW_TYPE_HINT_DIALOG</property>
+    <property name="skip_taskbar_hint">True</property>
+    <property name="has_separator">False</property>
+    <property name="message_type">GTK_MESSAGE_QUESTION</property>
+    <property name="buttons">GTK_BUTTONS_YES_NO</property>
+    <property name="text" translatable="yes">It appears that your graphics driver does not support the necessary extensions to use this tool.  Do you want to use your graphics driver vendor's tool instead?</property>
+    <property name="image"/>
+    <child internal-child="vbox">
+      <object class="GtkVBox" id="dialog-vbox2">
+        <property name="visible">True</property>
+        <property name="spacing">2</property>
+        <child internal-child="action_area">
+          <object class="GtkHButtonBox" id="dialog-action_area2">
+            <property name="visible">True</property>
+            <property name="layout_style">GTK_BUTTONBOX_END</property>
+          </object>
+          <packing>
+            <property name="expand">False</property>
+            <property name="pack_type">GTK_PACK_END</property>
+          </packing>
+        </child>
+      </object>
+    </child>
+  </object>
   <object class="GtkListStore" id="liststore1">
     <columns>
       <!-- column-name item -->
Index: gnome-control-center-2.31.90/capplets/display/Makefile.am
===================================================================
--- gnome-control-center-2.31.90.orig/capplets/display/Makefile.am	2010-09-01 17:42:32.282617001 +1000
+++ gnome-control-center-2.31.90/capplets/display/Makefile.am	2010-09-01 17:42:34.192617001 +1000
@@ -2,6 +2,9 @@
 cappletname = display
 
 uidir = $(pkgdatadir)/ui
+nvidia_settings = /usr/bin/nvidia-settings
+amdcccle = /usr/bin/amdcccle
+
 dist_ui_DATA = display-capplet.ui
 
 bin_PROGRAMS = gnome-display-properties
@@ -57,6 +60,8 @@
              $(GNOMECC_CAPPLETS_CFLAGS) \
 	     -DSBINDIR="\"$(sbindir)\"" \
 	     -DUIDIR="\"$(uidir)\"" \
+	     -DNVIDIA_SETTINGS="\"$(nvidia_settings)\"" \
+	     -DAMDCCCLE="\"$(amdcccle)\"" \
 	     -DGNOMELOCALEDIR="\"$(datadir)/locale\"" \
 	     -DGNOMECC_DATA_DIR="\"$(pkgdatadir)\""
 
Index: gnome-control-center-2.31.90/capplets/display/xrandr-capplet.c
===================================================================
--- gnome-control-center-2.31.90.orig/capplets/display/xrandr-capplet.c	2010-09-01 17:42:24.862617002 +1000
+++ gnome-control-center-2.31.90/capplets/display/xrandr-capplet.c	2010-09-01 17:42:34.192617001 +1000
@@ -474,7 +474,6 @@
     return count;
 }
 
-#if 0
 static int
 count_all_outputs (GnomeRRConfig *config)
 {
@@ -485,7 +484,6 @@
 
     return i;
 }
-#endif
 
 /* Computes whether "Mirror Screens" (clone mode) is supported based on these criteria:
  *
@@ -2256,7 +2254,6 @@
     begin_version2_apply_configuration (app, gtk_widget_get_window (app->dialog), app->apply_button_clicked_timestamp);
 }
 
-#if 0
 /* Returns whether the graphics driver doesn't advertise RANDR 1.2 features, and just 1.0 */
 static gboolean
 driver_is_randr_10 (GnomeRRConfig *config)
@@ -2277,7 +2274,6 @@
 
     return (count_all_outputs (config) == 1 && strcmp (config->outputs[0]->name, "default") == 0);
 }
-#endif
 
 static void
 on_detect_displays (GtkWidget *widget, gpointer data)
@@ -2542,6 +2538,8 @@
     GtkBuilder *builder;
     GtkWidget *align;
     GError *error;
+    GtkWidget *active_dialog;
+    gchar* command=NULL;
 
     error = NULL;
     builder = gtk_builder_new ();
@@ -2650,12 +2648,25 @@
     g_signal_connect (app->apply_button, "clicked",
 		      G_CALLBACK (apply_button_clicked_cb), app);
 
+	active_dialog=app->dialog;
     on_screen_changed (app->screen, app);
 
+    if (driver_is_randr_10(app->current_configuration))
+    {
+        if (g_file_test(NVIDIA_SETTINGS,G_FILE_TEST_IS_EXECUTABLE))
+            command=NVIDIA_SETTINGS;
+        else if (g_file_test(AMDCCCLE,G_FILE_TEST_IS_EXECUTABLE))
+            command=AMDCCCLE;
+        if (command != NULL)
+        {
+            active_dialog = _gtk_builder_get_widget (builder, "legacy_question");
+        }
+    }
+
     g_object_unref (builder);
 
 restart:
-    switch (gtk_dialog_run (GTK_DIALOG (app->dialog)))
+    switch (gtk_dialog_run (GTK_DIALOG (active_dialog)))
     {
     default:
 	/* Fall Through */
@@ -2682,6 +2693,18 @@
 	make_default (app);
 	goto restart;
 	break;
+
+	/* Legacy dialog question posed, but user wants this tool */
+	case GTK_RESPONSE_NO:
+	gtk_widget_destroy(active_dialog);
+	active_dialog=app->dialog;
+	goto restart;
+	break;
+
+	/* Legacy dialog question posed, user wants vendor tool */
+	case GTK_RESPONSE_YES:
+            g_spawn_command_line_async (command, NULL);
+    break;
     }
 
     gtk_widget_destroy (app->dialog);
